compiler/packages/babel-plugin-react-compiler/src/Optimization/OutlineJsx.ts

```diff
@@ -0,0 +1,416 @@
+        case 'LoadGlobal': {
+          globals.set(lvalue.identifier.id, instr as LoadGlobalInstruction);
+          break;
+        }
+        case 'FunctionExpression': {
+          outlineJsxImpl(value.loweredFunc.func, outlinedFns);
+          break;
+        }
+
+        case 'JsxExpression': {
+          if (!state.children.has(lvalue.identifier.id)) {
+            processAndOutlineJSX(state, rewriteInstr);
+
+            state = {
+              jsx: [],
+              children: new Set(),
+            };
+          }
+          state.jsx.push(instr as JsxInstruction);
+          if (value.children) {
+            for (const child of value.children) {
+              state.children.add(child.identifier.id);
+            }
+          }
+          break;
+        }
+      }
+    }
+    processAndOutlineJSX(state, rewriteInstr);
+
+    if (rewriteInstr.size > 0) {
+      const newInstrs = [];
+      for (let i = 0; i < block.instructions.length; i++) {
+        // InstructionId's are one-indexed, so add one to account for them.
+        const id = i + 1;
+        if (rewriteInstr.has(id)) {
+          const instrs = rewriteInstr.get(id);
+          newInstrs.push(...instrs);
+        } else {
+          newInstrs.push(block.instructions[i]);
+        }
+      }
+      block.instructions = newInstrs;
+    }
+    deadCodeElimination(fn);
+  }
+}
+
+type OutlinedResult = {
+  instrs: Array<Instruction>;
+  fn: HIRFunction;
+};
+
+function process(
+  fn: HIRFunction,
+  jsx: Array<JsxInstruction>,
+  globals: LoadGlobalMap,
+): OutlinedResult | null {
+  /**
+   * In the future, add a check for backedge to outline jsx inside loops in a
+   * top level component. For now, only outline jsx in callbacks.
+   */
+  if (fn.fnType === 'Component') {
+    return null;
+  }
+
+  const props = collectProps(jsx);
+  if (!props) return null;
+
+  const outlinedTag = fn.env.generateGloballyUniqueIdentifierName(null).value;
+  const newInstrs = emitOutlinedJsx(fn.env, jsx, props, outlinedTag);
+  if (!newInstrs) return null;
+
+  const outlinedFn = emitOutlinedFn(fn.env, jsx, props, globals);
+  if (!outlinedFn) return null;
+  outlinedFn.id = outlinedTag;
+
+  return {instrs: newInstrs, fn: outlinedFn};
+}
+
+function collectProps(
+  instructions: Array<JsxInstruction>,
+): Array<JsxAttribute> | null {
+  const attributes: Array<JsxAttribute> = [];
+  const jsxIds = new Set(instructions.map(i => i.lvalue.identifier.id));
+  const seen: Set<string> = new Set();
+  for (const instr of instructions) {
+    const {value} = instr;
+
+    for (const at of value.props) {
+      if (at.kind === 'JsxSpreadAttribute') {
+        return null;
+      }
+
+      /*
+       * TODO(gsn): Handle attributes that have same value across
+       * the outlined jsx instructions.
+       */
+      if (seen.has(at.name)) {
+        return null;
+      }
+
+      if (at.kind === 'JsxAttribute') {
+        seen.add(at.name);
+        attributes.push(at);
+      }
```

Comment:
nit: use an exhaustive switch here? it's not a big deal though, realistically we aren't going to add new attribute kinds

Reply:
done